### PR TITLE
fix(decorator-strategy): check headers values as strings

### PR DESCRIPTION
**Pull request by fancyweb**
## Subject

I am targeting this branch, because it's a bug fix.

This project does not properly compare HttpFoundation headers values. The returned value can be a string, an array of strings or null. The following commit : https://github.com/symfony/symfony/commit/ef5ead00059fc321e7fda20c2dfdd4440585be26#diff-5f886132f28113d74a1a219e967e88fcR124 recently enforced the returned values (by casting them to strings). Thus, the strict comparisons against booleans does not work anymore.

The bug was not detected by the CI because the wrong bag is used for headers in the unit test. I updated it and I also added assertions for the non tested header `x-sonata-page-not-decorable`.

I think it's better to handle both cases (bool & string) to not force a `symfony/http-foundation` bump.

## Changelog

```markdown
### Fixed
- Fixed `DecoratorStrategy` compatibility with `symfony/http-foundation` >= 3.4.31
```

### CODE REVIEW

File: src/CmsManager/DecoratorStrategy.php [hunk 1]
```diff
@@ -57,12 +57,14 @@ public function isDecorable(Request $request, $requestType, Response $response)
             return false;
         }
 
-        if (true === $response->headers->get('x-sonata-page-not-decorable', false)) {
+        // TODO: remove true support when the symfony/http-foundation dependency is > 3.4.31
```
Comment (phansys) [inline src/CmsManager/DecoratorStrategy.php]:
```suggestion
        // TODO: drop support for `true` value when bumping symfony/http-foundation dependency to > 3.4.31
```

File: src/CmsManager/DecoratorStrategy.php [hunk 2]
```diff
@@ -57,12 +57,14 @@ public function isDecorable(Request $request, $requestType, Response $response)
...
+        if (\in_array($response->headers->get('x-sonata-page-not-decorable', '0'), [true, '1'], true)) {
             return false;
         }
 
+        // TODO: remove true support when the symfony/http-foundation dependency is > 3.4.31
```
Comment (phansys) [inline src/CmsManager/DecoratorStrategy.php]:
```suggestion
        // TODO: drop support for `true` value when bumping symfony/http-foundation dependency to > 3.4.31
```

File: tests/CmsManager/DecoratorStrategyTest.php [hunk 3]
```diff
@@ -15,13 +15,14 @@
...
 use Sonata\PageBundle\CmsManager\DecoratorStrategy;
-use Symfony\Component\HttpFoundation\ParameterBag;
 use Symfony\Component\HttpFoundation\Request;
 use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\HttpFoundation\ResponseHeaderBag;
 use Symfony\Component\HttpKernel\HttpKernelInterface;
 
 class DecoratorStrategyTest extends TestCase
 {
+    // TODO: remove headers boolean support when the symfony/http-foundation dependency is > 3.4.31
```
Comment (phansys) [inline tests/CmsManager/DecoratorStrategyTest.php]:
```suggestion
    // TODO: drop support for `true` value when bumping symfony/http-foundation dependency to > 3.4.31

```

**Comment by fancyweb** [comment]
@phansys Thanks for your suggestions. I could not apply them directly because I had to correct the `>` to `>=`.

File: src/CmsManager/DecoratorStrategy.php [hunk 4]
```diff
@@ -57,12 +57,14 @@ public function isDecorable(Request $request, $requestType, Response $response)
             return false;
         }
 
-        if (true === $response->headers->get('x-sonata-page-not-decorable', false)) {
+        // TODO: drop support for `true` value when bumping symfony/http-foundation dependency to >= 3.4.31
+        if (\in_array($response->headers->get('x-sonata-page-not-decorable', '0'), [true, '1'], true)) {
```
Comment (core23) [inline src/CmsManager/DecoratorStrategy.php]:
```suggestion
        if (true === (bool) $response->headers->get('x-sonata-page-not-decorable', false)) {
```
Is more understandable

Comment (fancyweb) [reply src/CmsManager/DecoratorStrategy.php]: More understandable but casting is less strict as it allows us to support "out of scope" HttpFoundation implementation (for example if userland extends the headers bag and break the contract) + the default value passed to `get` must be a string.

Comment (greg0ire) [reply src/CmsManager/DecoratorStrategy.php]: Why not use `getBoolean` here?

Comment (fancyweb) [reply src/CmsManager/DecoratorStrategy.php]: `getBoolean` doesn't exist in `ResponseHeaderBag`

Comment (greg0ire) [reply src/CmsManager/DecoratorStrategy.php]: Oh right, I got confused with `ParameterBag`, nevermind

Comment (fancyweb) [reply src/CmsManager/DecoratorStrategy.php]: It was a great suggestion actually, I was disappointed too 😅

**Comment by greg0ire** [comment]
Note that 2.8 will no longer be supported, even for security fixes, so a bump to 3.4.31 would really not be an issue to me.

**Comment by core23** [comment]
Thanks @fancyweb

**Pull request merged**